### ISSUE TITLE
How to decode generic attributes in Javascript?

### ISSUE THREAD

**Issue by swomack** [google/draco#99]
I couldn't find any way to do it. Is there any way to search for the `custom_id` in javascript decoder? I can get the number of vertex attributes but can't iterate over the attributes.

**Comment by ondys** [comment]
A javascript example of iterating over attributes and using `custom_id` is shown for example in this comment: https://github.com/google/draco/issues/67#issuecomment-283124761

**Comment by swomack** [comment]
@ondys 

I have notices that the compression ratio gets reduced quite a bit when i add custom vertex attributes (a float value for each vertices) using `GeometryAttribute::GENERIC`.  Is it normal?

**Comment by ondys** [comment]
That really depends. Compression is affected by two factors:
1. Compression settings - quantization vs lossless compression, prediction schemes, etc..  What settings did you use?
2. Type of data - if the data distribution over the vertices is not spatially collocated it's going to compress worse.

**Comment by swomack** [comment]
Compressing with Draco is changing the vertices order, but vertices order is important to me. So i planned to send the vertices orders as `GeometryAttribute::GENERIC` attribute. This attribute will contain `unsigned int` from `0` to the `number of vertices - 1`. 

I am using the following settings -  

position quantization bits = 24
color quantization bits = 14
compression level = 10
speed = 10


One more thing, if the number of vertices is large, then some of the values of generic attributes are getting changed. Like, for 3.5 millions of vertices, 9 generic attribute's value has been changed.

**Comment by ondys** [comment]
Yes, this kind of generic attribute is most likely not going to compress that well as every value is unique. You may try to force prediction scheme to be delta coding (`PREDICTION_DIFFERENCE`). You may also try using sequential encoding method which is order preserving (`MESH_SEQUENTIAL_ENCODING`). 

As for the settings, assuming you are using C++ api directly, then compression level is inverse of speed so for compression level 10 you should set speed to 0 (for both encoding and decoding). If you set speed to 10 you are essentially setting you compression level to 0.

Lastly, for the changed values of generic attribute. That should certainly not happen. Are you by any chance using Javascript decoder? In our current Javascript API we always convert all attribute values to float32 data type that may not be sufficient to hold all your vertex indices. If your problem is with C++ API, then it would be good to provide some sample case that would allow us to reproduce the problem.

**Comment by swomack** [comment]
This speed is actually set to 0. That was a typo. Sorry for that.

Yeah, I am using javascript decoder. I guess, using `MESH_SEQUENTIAL_ENCODING` is my best bet.

**Comment by ondys** [comment]
@swomack I've created a new internal issues for both exploring better ways for doing order preserving compression + adding support for integer attributes to our javascript API.

**Comment by swomack** [comment]
Thats great!! Hopefully we can see it in the next release! That would be a great help since i have already integrated it in my project.

**Comment by swomack** [comment]
@ondys One more thing. I have `2191615` unsigned int generic attributes. The ranges of that generic attribute's values are `0 - 2191614`. Now, from these attribute values, `9` values are missing if i use javascript decoder. Values of the range `2191606 - 2191614` are missing. Now, even if we convert the `unsigned int` attribute value to `float32` there should not be any overflow for these attribute values. What do you think?

**Comment by ondys** [comment]
yes this should be well within the precision limits of float32.  What is the decoded value on the points that originally had values 2191606 to 2191614 ?

**Comment by swomack** [comment]
The decoded values are - 235197, 233182, 233139, 233096, 233052, 233009, 232966, 232924, 232882.

I will continue testing and share my findings.

**Comment by swomack** [comment]
It seems like anything larger than 2191605 is getting changed. Any value from 0 - 2191605 is fine. Rest are missing.

**Comment by ondys** [comment]
when you decode it with the C++ api, is it still wrong?

If you want me to look into it, can you post a link to the models (source + .drc)?

**Comment by swomack** [comment]
@ondys 

Sorry for being late. Exactly same thing is happening for the c++ decoder API. Same values are missing. Also, the model is a step file and we parse it using our API to generate vertices and indices. So, i am afraid i cant share these things. You should be able to reproduce this issue by adding a dummy generic attribute with data 0 to 2191614.

**Comment by ondys** [comment]
Ok, I'll try to create a repro case on our end.

**Comment by swomack** [comment]
Hello @ondys 

Is there any update regarding this?

**Comment by ondys** [comment]
@swomack Sorry not yet. I'm planning to look at it this week.

**Comment by ondys** [comment]
@swomack I tried it on a sample model with about 9M vertices and what I have noticed that some vertices can disappear when they are either isolated or they are connected elusively to degenerate faces. Other than that, all ids have been properly preserved (C++ api).

If you want to check how many isolated vertices your model has, you can either put a breakpoint at the end of `CornerTable::Initialize()` method  (file: `mesh/corner_table.h` and look for variable `num_isolated_vertices_` or you can try to output the value to console.. something like this:
```
bool CornerTable::Initialize(
    const IndexTypeVector<FaceIndex, FaceType> &faces) {
  faces_ = faces;
  int num_vertices = -1;
  if (!ComputeOppositeCorners(&num_vertices))
    return false;
  if (!ComputeVertexCorners(num_vertices))
    return false;
  printf("num isolated vert = %d\n", (int)NumIsolatedVertices());
  return true;
}
```

In either case, please let me know if this is the case for you models. If not we can investigate further.
